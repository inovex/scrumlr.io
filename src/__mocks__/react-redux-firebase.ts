import { get } from 'lodash';
const mod: any = jest.genMockFromModule('react-redux-firebase');

let _state = Object.create(null);

function setState(state: object) {
  _state = { ...state };
}
mod.__setState = setState;

mod.__signInAnonymously = jest
  .fn()
  .mockReturnValue(Promise.resolve({ ..._state.auth }));
mod.__signOut = jest.fn().mockReturnValue(Promise.resolve());
mod.__updateProfile = jest.fn().mockReturnValue(Promise.resolve());

function auth() {
  return {
    currentUser: {
      updateProfile: mod.__updateProfile,
      ..._state.auth
    },
    signInAnonymously: mod.__signInAnonymously,
    signOut: mod.__signOut
  };
}

mod.__ref__push = jest.fn().mockReturnValue(Promise.resolve());

function ref() {
  return {
    push: mod.__ref__push
  };
}

mod.getFirebase = jest.fn().mockReturnValue({
  auth,
  ref
});

function getVal(state: any, path: string, defaultValue: any) {
  const selector = (path.substring(0, 1) === '/'
    ? path.substring(1)
    : path).replace('/', '.');
  return get(_state, selector, defaultValue);
}

mod.getVal = getVal;

mod.firebaseConnect = jest.fn((cmp: any) => {
  return () => cmp;
});

module.exports = mod;
