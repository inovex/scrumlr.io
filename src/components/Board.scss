@import "common";

.board {
  --n: 1;

  display:grid;
  grid-auto-flow: column;
  grid-auto-columns: calc(100%/min(var(--real-columns), var(--n)));
  overflow-x:auto;
  scroll-snap-type: x mandatory;

  height: 100vh;
}

@for $i from 1 through 12 {
  @media screen and (min-width: $i*$min-column-width) {
    .board {
      --n: #{$i};
    }
  }
}