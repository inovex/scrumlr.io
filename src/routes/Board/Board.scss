.board-page {
  outline: none;
  display: flex;
  flex-direction: column;
  min-height: 430px;
  overflow: hidden;
}

.board-page__column-view {
  flex: 1;
}

.phase-splash__animation-enter, .phase-splash__animation-appear {
  // FIXME adding scale on enter breaks carousel calculation
  //transform: scale(.1);
  opacity: .005 !important;
}

.phase-splash__animation-appear.phase-splash__animation-appear-active,
.phase-splash__animation-enter.phase-splash__animation-enter-active {
  transform: scale(1);
  opacity: .95 !important;
  transition: opacity 200ms ease-in, transform 300ms ease-in;
}

.phase-splash__animation-leave {
  transform: scale(1);
  opacity: .95 !important;
}

.phase-splash__animation-leave.phase-splash__animation-leave-active {
  transform: scale(4);
  opacity: .005 !important;
  transition: opacity 300ms ease-in, transform 300ms ease-in;
}
