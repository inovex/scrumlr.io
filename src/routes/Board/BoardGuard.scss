@import "constants/style";

.board-guard {
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
}

.board-guard__denied-link {
  color: $color-planning-pink;
  transition: all 0.08s ease-out;

  &:hover {
    filter: brightness(1.4);
    transform: scale(1.02);
  }
}

.board-guard__info {
  font-size: $text-size--large;
}

[theme="dark"] {
  .board-guard {
    background-color: $color-dark-one;
  }

  .board-guard__info {
    color: $color-white;
  }
}
