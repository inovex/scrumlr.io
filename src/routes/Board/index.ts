export * from "./Board";
export * from "./BoardGuard";
