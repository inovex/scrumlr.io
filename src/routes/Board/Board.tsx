import * as React from 'react';
import { RouteComponentProps } from 'react-router';
import { useFirestoreConnect, isLoaded, isEmpty } from 'react-redux-firebase';
import { ApplicationState } from 'store/ApplicationState';
import { useSelector } from 'react-redux';
import LoadingScreen from "components/LoadingScreen/LoadingScreen";
import BoardData from "../../types/BoardData";

export interface BoardProps extends RouteComponentProps<{id: string}> {}

function Board(props: BoardProps) {

    const boardId = props.match.params.id;

    useFirestoreConnect([
        {collection: 'boards', doc: boardId}
    ]);

    const board: BoardData = useSelector((state: ApplicationState) => state.firestore.data.boards && state.firestore.data.boards[boardId]);

    if (isLoaded(board)) {
        if (isEmpty(board)) {
            // PLACEHOLDER: BOARD NON EXISTENT
            return <div> Board non existent</div>;
        } else {
            // PLACEHOLDER: INSERT BOARD
            return <div>
                <h1>{board.topic} ({boardId})</h1>
                <span>Owner: {board.owner}</span>
                <br/>
                <span>Date: {board.date.toDate().toString()}</span>
            </div>;
        }
    } else {
        return <LoadingScreen/>;
    }
}

export default Board;