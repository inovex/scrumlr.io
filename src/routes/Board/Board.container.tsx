import { AnyAction, Dispatch } from 'redux';
import { isLoaded, getVal } from 'react-redux-firebase';
import * as Raven from 'raven-js';
import { debounce, difference } from 'lodash';

import { SETUP_COMPLETED } from '../../actions';
import { BoardProps } from './Board';
import {
  BoardConfig,
  FirebaseProp,
  StoreState,
  BoardCards,
  Card,
  Optional,
  PrivateBoardData,
  PublicBoardData
} from '../../types';
import { authController } from '../../controller/auth';
import { Crypto } from '../../util/crypto';

export const mapStateToProps = (
  state: StoreState,
  ownProps: BoardProps
): BoardProps => {
  const { app, fbState } = state;
  const firebase = ownProps.firebase as FirebaseProp;

  const boardSelector = `boards/${ownProps.match.params.id}/private`;
  const boardPrintUrl = `/print/${ownProps.match.params.id}`;
  const auth: firebase.User | any = firebase.auth().currentUser || {};

  const board: Optional<PrivateBoardData> = getVal(
    fbState,
    `data/${boardSelector}`,
    undefined
  );

  if (!isLoaded(board) || !board) {
    return { boardSelector, boardConfig: undefined, auth } as any;
  }

  const cards: BoardCards = board.cards || {};
  const boardConfig: BoardConfig = board.config;

  const users = isLoaded(board) ? board.users : {};

  const adminUsers = boardConfig.adminUsers || {};

  //There should be only one board admin check - here!
  const isBoardAdmin = isLoaded(board)
    ? auth.uid === boardConfig.creatorUid || adminUsers[auth.uid]
    : false;

  let focusedCard: Optional<Card> = undefined;
  if (boardConfig.focusedCardId) {
    focusedCard = cards[boardConfig.focusedCardId];
  }

  let timerExpiration: Optional<string> = undefined;
  if (boardConfig.timerExpiration) {
    timerExpiration = boardConfig.timerExpiration;
  }

  const publicBoardSelector = `boards/${ownProps.match.params.id}/public`;
  const publicBoard: Optional<PublicBoardData> = getVal(
    fbState,
    `data/${publicBoardSelector}`,
    undefined
  );

  let waitingUsers: {
    uid: string;
    name: string;
    image: string;
    time: Date;
  }[] = [];
  if (isLoaded(publicBoard) && publicBoard) {
    const { accessAuthorized, applicants } = publicBoard;

    if (applicants) {
      const waitingUserUids = difference(
        Object.keys(applicants),
        Object.keys(accessAuthorized || {})
      );
      waitingUsers = waitingUserUids
        .map(uid => ({
          uid,
          ...applicants[uid],
          time: new Date(applicants[uid].time)
        }))
        .sort((a, b) => {
          if (a.time > b.time) {
            return 1;
          } else if (a.time < b.time) {
            return -1;
          }
          return 0;
        });
    }
  }

  function acceptUser(uid: string, accept: boolean) {
    firebase
      .set(`${publicBoardSelector}/accessAuthorized/${uid}`, accept)
      .catch((err: any) => {
        Raven.captureMessage('unable to accept user access', {
          extra: {
            reason: err.message,
            uid: auth.uid,
            boardId: boardSelector,
            user: uid,
            accept
          }
        });
      });
  }

  // TODO cleanup symmetric key sync
  if (isBoardAdmin && users) {
    const keyStore = board.keyStore;

    if (keyStore) {
      const crypto = new Crypto();
      crypto.initKeypair().then(() => {
        const key = keyStore[crypto.getPublicKey()!!];

        if (key) {
          crypto.importSymmetricKey(key).then(() => {
            for (let uid in users) {
              if (keyStore && !keyStore[users[uid].publicKey!!]) {
                crypto
                  .exportSymmetricKey(users[uid].publicKey!!)
                  .then(exportedKey => {
                    firebase.set(
                      `${boardSelector}/keyStore/${users[uid].publicKey}`,
                      exportedKey
                    );
                  });
              }
            }
          });
        }
      });
    }
  }

  function onToggleShowAuthor() {
    firebase
      .update(`${boardSelector}/config`, {
        showAuthor: !boardConfig.showAuthor
      })
      .catch((err: Error) => {
        Raven.captureMessage('Could not toggle show author state', {
          extra: { reason: err.message, uid: auth.uid, boardId: boardSelector }
        });
      });
  }

  function onToggleReadyState() {
    firebase
      .update(`${boardSelector}/users/${auth.uid}`, {
        ready: !board!!.users[auth.uid].ready
      })
      .catch((err: Error) => {
        Raven.captureMessage('Could not toggle user state', {
          extra: { reason: err.message, uid: auth.uid, boardId: boardSelector }
        });
      });
  }

  function onFocusCard(cardId: string) {
    const { focusedCardId } = boardConfig;
    firebase
      .set(
        `${boardSelector}/config/focusedCardId`,
        focusedCardId !== cardId ? cardId : null
      )
      .catch((err: any) => {
        Raven.captureMessage('Could not focus card', {
          extra: {
            reason: err.message,
            uid: auth.uid,
            boardId: boardSelector,
            focusedCardId
          }
        });
      });
  }

  const username = auth.displayName || undefined;
  const email = auth.email || undefined;
  const isAnonymous = auth.isAnonymous;

  const onChangeUsername = debounce((username: string) => {
    const user = firebase.auth().currentUser;
    if (user) {
      user
        .updateProfile({
          displayName: username,
          photoURL: user.photoURL
        })
        .then(() => {
          firebase.update(`${boardSelector}/users/${auth.uid}`, {
            name: username
          });
        });
    }
  }, 2000);

  const onChangeBoardName = debounce((boardName: string) => {
    const { focusedCardId } = boardConfig;
    firebase
      .set(`${boardSelector}/config/name`, boardName)
      .catch((err: any) => {
        Raven.captureMessage('Could not set boardname', {
          extra: {
            reason: err.message,
            uid: auth.uid,
            boardId: boardSelector,
            focusedCardId
          }
        });
      });
  }, 2000);

  const onDeleteTimer = () => {
    firebase
      .set(`${boardSelector}/config/timerExpiration`, null)
      .catch((err: any) => {
        Raven.captureMessage('Unable to delete timer', {
          extra: {
            reason: err.message,
            uid: auth.uid,
            boardId: boardSelector
          }
        });
      });
  };

  function onSwitchPhaseIndex(delta: number) {
    firebase
      .ref(`${boardSelector}/config/guidedPhase`)
      .transaction((phase: number) => phase + delta)
      .catch((err: Error) => {
        Raven.captureMessage('Could not switch phase index', {
          extra: { reason: err.message, uid: auth.uid, boardId: boardSelector }
        });
      });

    const updateUsers = {};
    Object.keys(board!!.users).forEach(uid => {
      updateUsers[`${uid}/ready`] = false;
    });

    console.log(updateUsers);
    firebase
      .ref(`${boardSelector}/users`)
      .update(updateUsers)
      .catch((err: Error) => {
        Raven.captureMessage('Could reset ready status for users', {
          extra: { reason: err.message, boardId: boardSelector }
        });
      });

    // Always remove card selection if new phase is entered.
    if (boardConfig.focusedCardId) {
      onFocusCard(boardConfig.focusedCardId);
    }
  }

  return {
    ...ownProps,
    cards,
    boardSelector,
    boardConfig,
    boardPrintUrl,
    users,
    focusedCard,
    isBoardAdmin,
    username,
    email,
    timerExpiration,
    isAnonymous,
    uid: auth.uid,
    onToggleReadyState,
    onFocusCard,
    onSwitchPhaseIndex,
    isShowAuthor: Boolean(boardConfig.showAuthor),
    onSignOut: authController(firebase).signOut,
    onChangeBoardName,
    onChangeUsername,
    onDeleteTimer,
    onToggleShowAuthor,
    onRegisterCurrentUser: () => null, // will be filled in mergeProps
    waitingUsers,
    acceptUser,
    ...app,

    // other props, only used in mergeProps
    auth
  };
};

export function mapDispatchToProps(dispatch: Dispatch<AnyAction>) {
  return {
    dispatch
  };
}

export function mergeProps(
  stateProps: BoardProps,
  { dispatch }: { dispatch: Dispatch<any> },
  ownProps: BoardProps
): BoardProps {
  function onRegisterCurrentUser() {
    dispatch({ type: SETUP_COMPLETED });
  }

  return {
    ...ownProps,
    ...stateProps,
    onRegisterCurrentUser
  };
}
