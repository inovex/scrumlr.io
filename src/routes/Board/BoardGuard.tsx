import {useEffect} from "react";
import store from "store";
import {ActionFactory} from "store/action";
import {RouteComponentProps} from "react-router";
import {ApplicationState} from "types/store";
import {useSelector} from "react-redux";
import LoadingIndicator from "components/LoadingIndicator/LoadingIndicator";
import Board from "./Board";
import "./BoardGuard.scss";

export type BoardGuardProps = RouteComponentProps<{id: string}>;

const BoardGuard = (props: BoardGuardProps) => {
  const boardStatus = useSelector((state: ApplicationState) => state.board.status);

  useEffect(() => {
    const boardId = props.match.params.id;
    store.dispatch(ActionFactory.joinBoard(boardId));

    return () => {
      store.dispatch(ActionFactory.leaveBoard());
    };
  }, [props.match.params.id]);

  if (boardStatus === "accepted" || boardStatus === "ready") {
    return <Board />;
  }
  if (boardStatus === "rejected") {
    return (
      <div className="board-guard">
        <p className="board-guard__info">You have been rejected.</p>
        <a href="/" className="board-guard__denied-link">
          Return to homepage
        </a>
      </div>
    );
  }
  return (
    <div className="board-guard">
      <LoadingIndicator />
      <p className="board-guard__info">Waiting for approval.</p>
    </div>
  );
};

export default BoardGuard;
