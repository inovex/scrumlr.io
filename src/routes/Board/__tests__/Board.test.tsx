import * as React from 'react';
import Board, {BoardProps} from '../Board';
import {render} from '@testing-library/react';
import {useFirestoreConnect, isLoaded, isEmpty} from 'react-redux-firebase';
import {useSelector} from 'react-redux';

jest.mock('react-redux-firebase');
jest.mock('react-redux');


describe('Board', () => {

    const mockProps: BoardProps = {
        match: {
            params: {id: '-foobar'},
            isExact: true,
            path: '/board/:id',
            url: '/board/-foobar'
        },
        location: null as any,
        history: null as any
    };

    describe('should render correctly', () => {

        const mockBoard = {
            topic: 'Hello World',
            owner: 'Admin',
            date: {
                toDate: () => 'date xyz'
            }
        };

        (useFirestoreConnect as jest.Mock).mockImplementation();
        (useSelector as jest.Mock).mockImplementation();

        test('on isLoaded == false', () => {
            (isLoaded as unknown as jest.Mock).mockReturnValueOnce(false);
            const {container} = render(<Board {...mockProps}/>);
            expect(container.firstChild).toMatchSnapshot();
        });

        test('on isLoaded == true and isEmpty == true', () => {
            (isLoaded as unknown as jest.Mock).mockReturnValueOnce(true);
            (isEmpty as unknown as jest.Mock).mockReturnValueOnce(true);
            const {container} = render(<Board {...mockProps}/>);
            expect(container.firstChild).toMatchSnapshot();
        });

        test('on isLoaded == true and isEmpty == false', () => {
            (isLoaded as unknown as jest.Mock).mockReturnValueOnce(true);
            (isEmpty as unknown as jest.Mock).mockReturnValueOnce(false);
            (useSelector as jest.Mock).mockImplementation().mockReturnValueOnce(mockBoard);
            const {container} = render(<Board {...mockProps}/>);
            expect(container.firstChild).toMatchSnapshot();
        });

    });

    test('should call useFirestoreConnect with correct queriesConfig', () => {
        const mockQueriesConfig = [{collection: 'boards', doc: '-foobar'}];
        const useFirestoreConnectMock = (useFirestoreConnect as jest.Mock).mockImplementation();
        (useSelector as jest.Mock).mockImplementation();
        render(<Board {...mockProps}/>);
        expect(useFirestoreConnectMock).toBeCalledWith(mockQueriesConfig);
    });



});