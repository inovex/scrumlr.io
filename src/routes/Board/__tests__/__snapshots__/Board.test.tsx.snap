// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board should render correctly on isLoaded == false 1`] = `
<div
  class="loading-screen"
>
  <div
    class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
    role="progressbar"
    style="width: 40px; height: 40px;"
  >
    <svg
      class="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</div>
`;

exports[`Board should render correctly on isLoaded == true and isEmpty == false 1`] = `
<div>
  <h1>
    Hello World
     (
    -foobar
    )
  </h1>
  <span>
    Owner: 
    Admin
  </span>
  <br />
  <span>
    Date: 
    date xyz
  </span>
</div>
`;

exports[`Board should render correctly on isLoaded == true and isEmpty == true 1`] = `
<div>
   Board non existent
</div>
`;
